mesh/mesh.go

```diff
@@ -241,21 +257,34 @@ func (m *Mesh) ExtractUniqueOrderedTransactions(l *types.Layer) []*Transaction {
 	return MergeDoubles(txs)
 }
 
-func (m *Mesh) PushTransactions(oldBase types.LayerID, newBase types.LayerID) {
+func (m *Mesh) PushTransactions(oldBase, newBase types.LayerID) {
 	for i := oldBase; i < newBase; i++ {
-
 		l, err := m.GetLayer(i)
 		if err != nil || l == nil {
 			m.Error("could not get layer %v  !!!!!!!!!!!!!!!! %v ", i, err) //todo handle error
 			return
 		}
 
-		merged := m.ExtractUniqueOrderedTransactions(l)
-		x, err := m.ApplyTransactions(types.LayerID(i), merged)
+		validBlockTxs, invalidBlockTxs := m.ExtractUniqueOrderedTransactions(l)
+		numAppliedTxs, err := m.ApplyTransactions(i, validBlockTxs)
 		if err != nil {
-			m.Log.Error("cannot apply transactions %v", err)
+			m.With().Error("cannot apply transactions", log.Err(err))
 		}
-		m.Log.Info("applied %v transactions in new pbase is %d apply result was %d", len(merged), newBase, x)
+		if err := m.removeFromMeshTxs(validBlockTxs, invalidBlockTxs, i); err != nil {
+			m.With().Error("failed to remove from meshTxs", log.Err(err))
+		}
+		if m.txMempool != nil {
```

Comment:
In practice, only in tests.

Since this value is set using a setter and not the constructor (for reasons of circular dependency), I can't guarantee that it's there, so this code is safer.